### ISSUE TITLE
 BASE_LR computed on IMS_PER_BATCH instead of Effective_Batch_Size. Confusion as to whether this is correct or not.

### ISSUE THREAD

**Issue by kevinj22** [roytseng-tw/Detectron.pytorch#113]
Hello,

The following is my code to start training:

python train_net_step.py --dataset coco2017 --cfg ../configs/baselines/e2e_mask_rcnn_R-50-FPN_1x.yaml --bs 2 --nw 4 --iter_size 8 

I use a batchsize of 2 with a iter_size of 8 to make the effective batch size 16. The automatic parameter update changes the base learning rate from 0.02 to 0.0025.

i.e 

Adjust BASE_LR linearly according to batch_size change:
    BASE_LR: 0.02 --> 0.0025

I tried manually setting the learning rate using the --lr 0.02 command but the printout states the learning rate is still 0.0025 when I resume training.

Is this correct? Shouldn't the base learning rate be 0.02 as the effective batch size is 16 as the update takes place after 8 iterations? Is this set because the gradients are accumulated as lr*(batch of 2) + next 8 iters?

Thanks,

Kevin

**Comment by roytseng-tw** [comment]
Yes, you are right. Gradients are accumulated in the period of iter_size, so the lr should be iter_size smaller to make the updates correct.

**Comment by kevinj22** [comment]
So this code should be adjusted based on effective batch size then?

```
### Adjust learning based on batch size change linearly
    # For iter_size > 1, gradients are `accumulated`, so lr is scaled based
    # on batch_size instead of effective_batch_size
    old_base_lr = cfg.SOLVER.BASE_LR
    cfg.SOLVER.BASE_LR *= args.batch_size / original_batch_size
    print('Adjust BASE_LR linearly according to batch_size change:\n'
          '    BASE_LR: {} --> {}'.format(old_base_lr, cfg.SOLVER.BASE_LR))
```

**Comment by roytseng-tw** [comment]
No. The code is correct.

**Comment by kevinj22** [comment]
Lets say we change the batch size to 2 and the iter size to 8, this changes the lr to 0.0025.

If we wait for 8 iters to pass we have an update of batch size 16. There's two way I can see of doing this update:

1. Sum of (lr * each batch of 2)
2. lr * (sum of all batches)

In both instances the lr, which should be equal to 0.02 as the effective batch size is 16, is still 0.0025.

What do I have wrong?

**Comment by roytseng-tw** [comment]
Learning rate is only applied once every iter_size iterations, on the update. Hence, the code is correct.
